python/py-mofka-client.cpp

```diff
@@ -3,43 +3,261 @@
+        if(__stride__ != (__buf_info__).strides[i])   \
+            throw mofka::Exception("MOFKA_ERR_NONCONTIG");  \
+        __stride__ *= (__buf_info__).shape[i];        \
+    }                                                 \
+} while(0)
+
+#define CHECK_BUFFER_IS_WRITABLE(__buf_info__) do { \
+    if((__buf_info__).readonly)                     \
+        throw mofka::Exception("MOFKA_ERR_READONLY");     \
+} while(0)
+
+
+template <typename DataType>
+auto data_helper(const DataType& data){
+    auto data_info = get_buffer_info(data);
+    CHECK_BUFFER_IS_CONTIGUOUS(data_info);
+    return mofka::Data(data_info.ptr, data_info.size);
+}
+
+template<typename DataType>
+static auto data_helper(const std::vector<DataType>& buffers) {
+    std::vector<mofka::Data::Segment> segments;
+    for (auto buff : buffers){
+        auto buff_info = get_buffer_info(buff);
+        CHECK_BUFFER_IS_CONTIGUOUS(buff_info);
+        CHECK_BUFFER_IS_WRITABLE(buff_info);
+        segments.push_back(mofka::Data::Segment(buff_info.ptr, 
+                                                buff_info.size));
+    }
+    return mofka::Data(segments);
+}
+
+static auto metadata_helper(const py::dict metadata) {
+    py::module_ json = py::module_::import("json");
+    std::string str_metadata = json.attr("dumps")(metadata).cast<std::string>();
+    return mofka::Metadata(str_metadata);
+}
+
+static auto data_broker_helper(const std::function<py::buffer(py::dict, mofka::DataDescriptor descriptor)> &broker,
+                                 const mofka::Metadata metadata = mofka::Metadata{"{\"type\":\"default\"}"}, 
+                                 const mofka::DataDescriptor data_descriptor = mofka::DataDescriptor()) {
+    return [broker](mofka::Metadata metadata, 
+                     mofka::DataDescriptor data_descriptor) -> mofka::Data {
+        py::module_ json = py::module_::import("json");
+        py::dict d_metadata = json.attr("loads")(metadata.string());
+        py::buffer data_buffer = broker(d_metadata, data_descriptor).cast<py::buffer>();
+        return data_helper(data_buffer);
+    };
+}
+
+static auto data_selector_helper(const std::function<mofka::DataDescriptor(py::dict, mofka::DataDescriptor descriptor)> &selector,
+                                 const mofka::Metadata metadata = mofka::Metadata{"{\"type\":\"default\"}"}, 
+                                 const mofka::DataDescriptor data_descriptor = mofka::DataDescriptor()) {
+    return  [selector](mofka::Metadata metadata, 
+                       mofka::DataDescriptor data_descriptor) -> mofka::DataDescriptor {
+        py::module_ json = py::module_::import("json");
+        py::dict d_metadata = json.attr("loads")(metadata.string());
+        return selector(d_metadata, data_descriptor);
+    };
+
+}
+           
+
+std::string stringify(const rapidjson::Value& v) {
+	if (v.IsString())
+		return { v.GetString(), v.GetStringLength() };
+	else {
+        rapidjson::StringBuffer strbuf;
```

Comment:
Almost: you can remove the "if" part because the "else" part is going to take care of the string case (also in the string case you would still need to add quotes, which the rapidjson Writer does for you). You should also replace strbuf.m_str with `std::move(strbuf).m_str` or `std::move(strbuf.m_str)`, otherwise the C++ compiler will create a copy of m_str instead of moving it.